From 098ced6bbc573f009217f3b191556038319d76fa Mon Sep 17 00:00:00 2001
From: MrLGamer <mrlgam3r@gmail.com>
Date: Mon, 21 Sep 2026 20:07:41 +0200
Subject: [PATCH 09/10] ntoskrnl.exe: Add a call wrapper for
 PsGetContextThread().

Add the KPROCESSOR_MODE argument to PsGetContextThread and keep the
existing context retrieval for both user-mode and kernel-mode requests.

Use an explicit call instruction in the x86-64 entry point. NIKKE's ACE
driver inspects this instruction to locate the context retrieval helper.

The wrapper assembly is adapted from NelloKudo
---
 dlls/ntoskrnl.exe/ntoskrnl.c        | 27 ++++++++++++++++++++++-----
 dlls/ntoskrnl.exe/ntoskrnl.exe.spec |  2 +-
 include/ddk/ntddk.h                 |  1 +
 3 files changed, 24 insertions(+), 6 deletions(-)

diff --git a/dlls/ntoskrnl.exe/ntoskrnl.c b/dlls/ntoskrnl.exe/ntoskrnl.c
index b8a7ea8..10127a8 100644
--- a/dlls/ntoskrnl.exe/ntoskrnl.c
+++ b/dlls/ntoskrnl.exe/ntoskrnl.c
@@ -2855,15 +2855,12 @@ HANDLE WINAPI PsGetThreadProcessId( PETHREAD thread )
     return thread->kthread.id.UniqueProcess;
 }
 
-/*********************************************************************
- *           PsGetContextThread    (NTOSKRNL.@)
- */
-NTSTATUS WINAPI PsGetContextThread(PETHREAD thread, CONTEXT *context)
+NTSTATUS WINAPI get_context_thread( PETHREAD thread, CONTEXT *context, KPROCESSOR_MODE mode )
 {
     NTSTATUS status;
     HANDLE h;
 
-    TRACE("%p %p\n", thread, context);
+    TRACE("%p %p %u\n", thread, context, mode);
 
     if ((status = ObOpenObjectByPointer(thread, 0, NULL, THREAD_ALL_ACCESS, NULL, KernelMode, &h)))
     {
@@ -2877,6 +2874,26 @@ NTSTATUS WINAPI PsGetContextThread(PETHREAD thread, CONTEXT *context)
     return status;
 }
 
+/*********************************************************************
+ *           PsGetContextThread    (NTOSKRNL.@)
+ */
+#ifdef __x86_64__
+__ASM_GLOBAL_FUNC( PsGetContextThread,
+                   "subq $0x28,%rsp\n\t"
+                   __ASM_SEH(".seh_stackalloc 0x28\n\t")
+                   __ASM_SEH(".seh_endprologue\n\t")
+                   __ASM_CFI(".cfi_adjust_cfa_offset 0x28\n\t")
+                   "call " __ASM_NAME("get_context_thread") "\n\t"
+                   "addq $0x28,%rsp\n\t"
+                   __ASM_CFI(".cfi_adjust_cfa_offset -0x28\n\t")
+                   "ret" )
+#else
+NTSTATUS WINAPI PsGetContextThread( PETHREAD thread, CONTEXT *context, KPROCESSOR_MODE mode )
+{
+    return get_context_thread( thread, context, mode );
+}
+#endif
+
 /***********************************************************************
  *           KeInsertQueue   (NTOSKRNL.EXE.@)
  */
diff --git a/dlls/ntoskrnl.exe/ntoskrnl.exe.spec b/dlls/ntoskrnl.exe/ntoskrnl.exe.spec
index 5674781..886f367 100644
--- a/dlls/ntoskrnl.exe/ntoskrnl.exe.spec
+++ b/dlls/ntoskrnl.exe/ntoskrnl.exe.spec
@@ -912,7 +912,7 @@
 @ stdcall PsDereferencePrimaryToken(ptr) ObDereferenceObject
 @ stub PsDisableImpersonation
 @ stub PsEstablishWin32Callouts
-@ stdcall PsGetContextThread(ptr ptr)
+@ stdcall PsGetContextThread(ptr ptr long)
 @ stdcall PsGetCurrentProcess() IoGetCurrentProcess
 @ stdcall PsGetCurrentProcessId()
 @ stdcall PsGetCurrentProcessSessionId()
diff --git a/include/ddk/ntddk.h b/include/ddk/ntddk.h
index 887d174..b80691e 100644
--- a/include/ddk/ntddk.h
+++ b/include/ddk/ntddk.h
@@ -288,6 +288,7 @@ void      WINAPI DECLSPEC_NORETURN KeBugCheck(ULONG);
 NTSTATUS  WINAPI KeExpandKernelStackAndCallout(PEXPAND_STACK_CALLOUT,void*,SIZE_T);
 void      WINAPI KeSetTargetProcessorDpc(PRKDPC,CCHAR);
 BOOLEAN   WINAPI MmIsAddressValid(void *);
+NTSTATUS  WINAPI PsGetContextThread(PETHREAD,CONTEXT*,KPROCESSOR_MODE);
 HANDLE    WINAPI PsGetProcessId(PEPROCESS);
 void *    WINAPI PsGetProcessSectionBaseAddress(PEPROCESS);
 HANDLE    WINAPI PsGetThreadId(PETHREAD);
-- 
2.53.0.697.g625c4fb2daa

